memiavl/tree_test.go

```diff
@@ -107,6 +118,133 @@ func TestNewKey(t *testing.T) {
 }
 
 func TestEmptyTree(t *testing.T) {
-	tree := New()
+	tree, err := New(DefaultPathToWAL)
+	require.NoError(t, err)
+
+	defer os.RemoveAll(DefaultPathToWAL)
+	defer tree.bwal.Close()
+
 	require.Equal(t, emptyHash, tree.RootHash())
 }
+
+func TestWAL(t *testing.T) {
+	tree, err := NewEmptyTree(0, DefaultPathToWAL)
+	require.NoError(t, err)
+
+	defer os.RemoveAll(DefaultPathToWAL)
+	defer tree.bwal.Close()
+
+	_, version, err := tree.ApplyChangeSet(DefaultChanges, true)
+	require.NoError(t, err)
+
+	// get data from WAL
+	data, err := tree.bwal.Read(uint64(version))
+	require.NoError(t, err)
+
+	expectedData := []byte{}
+	for _, i := range DefaultChangesBz {
+		expectedData = append(expectedData, i...)
+	}
+
+	require.Equal(t, expectedData, data)
+}
+
+func TestReplayWAL(t *testing.T) {
+	t.Cleanup(removeDefaultWal) // removes default wal file from filesystem
+
+	tree, err := NewEmptyTree(0, DefaultPathToWAL)
+	require.NoError(t, err)
+	secondTreeWALPath := "test_wal"
+
+	defer os.RemoveAll(secondTreeWALPath)
+
+	for _, cs := range ChangeSets {
+		_, _, err := tree.ApplyChangeSet(cs, false)
+		require.NoError(t, err)
+	}
+	tree.RootHash()
+
+	// replay WAL
+	tree2, err := NewEmptyTree(0, secondTreeWALPath)
+	require.NoError(t, err)
+	defer tree2.bwal.Close()
+
+	randomWalPath := "random_wal_path"
+	defer os.RemoveAll(randomWalPath)
+
+	tests := map[string]struct {
+		untilVersion uint64
+		walPath      string
+		tree         *Tree
+		expectingErr bool
+	}{
+		"valid: replay all versions": {
+			untilVersion: 0,
+			tree:         tree2,
+			walPath:      DefaultPathToWAL,
+		},
+		"invalid: untilVersion too high": {
+			untilVersion: 100,
+			tree:         tree2,
+			walPath:      DefaultPathToWAL,
+			expectingErr: true,
+		},
+		"invalid: tree already up to date": {
+			untilVersion: 0,
+			tree:         tree,
+			walPath:      DefaultPathToWAL,
+			expectingErr: true,
+		},
+		"random wal path: will create empty new wal: should fail": {
+			untilVersion: 0,
+			tree:         tree2,
+			walPath:      randomWalPath,
+			expectingErr: true,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
```

Comment:
might need mv the Cleanup before each test

Reply:
these subtests do not create any new wals, so I do not think we should